repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/services/datastore/RemoteDatastoreServiceTest.groovy

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package features.services.datastore
+
+import framework.*
+import framework.category.Slow
+import org.apache.commons.io.filefilter.WildcardFileFilter
+import org.junit.experimental.categories.Category
+import org.linkedin.util.clock.SystemClock
+import org.rackspace.deproxy.Deproxy
+import spock.lang.Shared
+import spock.lang.Specification
+
+import java.nio.file.Files
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+import static javax.servlet.http.HttpServletResponse.SC_OK
+import static javax.servlet.http.HttpServletResponse.SC_REQUEST_ENTITY_TOO_LARGE
+import static org.linkedin.groovy.util.concurrent.GroovyConcurrentUtils.waitForCondition
+
+@Category(Slow.class)
+class RemoteDatastoreServiceTest extends Specification {
+    @Shared
+    def repose1
+
+    @Shared
+    def repose2
+
+    @Shared
+    def remote
+
+    @Shared
+    def deproxy
+
+    @Shared
+    def repose1Port
+
+    @Shared
+    def repose2Port
+
+    @Shared
+    def reposeOneEndpoint
+
+    @Shared
+    def reposeTwoEndpoint
+
+    @Shared
+    def remoteDatastoreEndpoint
+
+    def setupSpec() {
+        repose1Port = PortFinder.instance.getNextOpenPort()
+        repose2Port = PortFinder.instance.getNextOpenPort()
+        def remotePort = PortFinder.instance.getNextOpenPort()
+        def datastorePort = PortFinder.instance.getNextOpenPort()
+        def targetPort = PortFinder.instance.getNextOpenPort()
+
+        reposeOneEndpoint = "http://localhost:$repose1Port"
+        reposeTwoEndpoint = "http://localhost:$repose2Port"
+        remoteDatastoreEndpoint = "http://localhost:$datastorePort"
+
+        deproxy = new Deproxy()
+        deproxy.addEndpoint(targetPort)
+
+        def launcherAndLog1 = startRepose('repose1', repose1Port, targetPort, datastorePort, true)
+        repose1 = launcherAndLog1.ReposeValveLauncher
+
+        def launcherAndLog2 = startRepose('repose2', repose2Port, targetPort, datastorePort, true)
+        repose2 = launcherAndLog2.ReposeValveLauncher
+
+        def launcherAndLogR = startRepose('remote', remotePort, targetPort, datastorePort, false)
+        remote = launcherAndLogR.ReposeValveLauncher
+
+        waitUntilReadyToServiceRequests(launcherAndLog1.ReposeLogSearch)
+        waitUntilReadyToServiceRequests(launcherAndLog2.ReposeLogSearch)
+        waitUntilReadyToServiceRequests(launcherAndLogR.ReposeLogSearch)
+    }
+
+    def cleanupSpec() {
+        deproxy?.shutdown()
+        repose1?.stop()
+        repose2?.stop()
+        remote?.stop()
+    }
+
+    static def startRepose(String subName, int reposePort, int targetPort, int datastorePort, boolean client) {
+        def testProperties = new TestProperties(RemoteDatastoreServiceTest.canonicalName.replace('.', '/') + "/$subName")
```

Comment:
A Java [Path](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html), probably. I basically never want to have to worry about the environment I'm running on, and `Path` lets me avoid that to a large extent. Of course, that would require a bit of rework because everything is a String right now.